source/agora/consensus/data/Params.d

```diff
@@ -69,7 +69,9 @@ public immutable class ConsensusParams
             validator_cycle = cycle length for a validator
             max_quorum_nodes = max nodes to include in autogenerated quorum set
             quorum_threshold = the threshold to use for the generated quorums
-            genesis_start_time = unix timestamp of the Genesis block
+            block_timestamp_tolerance_secs = new block timestamp has to be greater
+                than previous block timestamp, but less than
```

Comment:
Well we can't really have a limit to the upper bound. If the rule is it has to be less than `current time + block_timestamp_tolerance_secs` then that would mean if the nodes are late to nominate they would be stuck forever?

Reply:
1. if we don't have an upper bound then nodes can agree on block timestamp that is 100 years into the future
2. what do you mean by 'if the nodes are late to nominate'? If a node clock is too far forward compared to other blocks, then yes it cannot nominate his own candidates, but it will re-nominate other node candidates

Reply:
see https://github.com/bpfkorea/agora/pull/1437#discussion_r546591387

if you mean the node's current clock is too back in the past, so it will not accept anything: I think we already have some assumption about how close the clocks of the nodes have to be together. Why was the original clock synchronization algorithm added? Where else do we need the clocks to be synchronized and what happen there if they are not?

Reply:
> if we don't have an upper bound then nodes can agree on block timestamp that is 100 years into the future

We still use the median value of the network time. So this would only fail if the majority of nodes were conspiring to do this.

In case of network outages (think an hour of downtime) we'll have to create a block once the network issues are resolved. But this value here might prevent us from doing that unless `block_timestamp_tolerance_secs` is set to a really high value.

Reply:
> > if we don't have an upper bound then nodes can agree on block timestamp that is 100 years into the future
> 
> We still use the median value of the network time. So this would only fail if the majority of nodes were conspiring to do this.
> 
> In case of network outages (think an hour of downtime) we'll have to create a block once the network issues are resolved. But this value here might prevent us from doing that unless `block_timestamp_tolerance_secs` is set to a really high value.

1. if an attacker modifies the agora code and nominates block with 100 years in the future AND there are no upper limit, then his nomination might win and we will have blocks with incorrect timestamp

2. I don't see any problems with network outages, lets say the last block timestamp was: 
1111 and that was the correct current time
the network outage last until 2222 and suppose we should have created 2 blocks
any block timestamp will be fine between 1111 and 2222 + tolerance for those 2 blocks

Reply:
the criterea for good block timestamp is:
greater then previous block timestamp AND 
less then the current network time + tolerance

Reply:
@ferencdg : There is a difference between "consensus parameters" and "node parameters".
A consensus parameter is something that the node cannot break without having a visible effect on the network, while a node parameter is something that can be broken unseen. Think about the way nodes deal with transactions they receive: A node could choose to ignore every transaction which has an odd hash, and it would behave according to the consensus protocol for the rest of the world.
On the other hand, if we have a "hash filter" on the block, and node can only elect odd hash for odd block height and even hash for even height, then it becomes a consensus parameter.

What you are describing here is a node parameter. A node should absolutely have a way to configure how far into a future it accepts to nominate a block. But it cannot be a hard, consensus limit, as @AndrejMitrovic noted.